platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingArtifactVariantSelector.java

```diff
@@ -80,78 +80,27 @@ private ResolvedArtifactSet doSelect(
         boolean allowNoMatchingVariants
     ) {
         AttributeMatcher matcher = attributeSchemaServices.getMatcher(consumerSchema, producer.getProducerSchema());
-        ImmutableAttributes componentRequested = attributesFactory.concat(requestAttributes, producer.getOverriddenAttributes());
-        final List<ResolvedVariant> variants = producer.getCandidates();
-
-        List<? extends ResolvedVariant> matches = matcher.matchMultipleCandidates(variants, componentRequested);
-        if (matches.size() == 1) {
-            return matches.get(0).getArtifacts();
-        } else if (matches.size() > 1) {
-            throw failureProcessor.ambiguousArtifactsFailure(matcher, producer, componentRequested, matches);
-        }
-
-        // We found no matches. Attempt to construct artifact transform chains which produce matching variants.
-        List<TransformedVariant> transformedVariants = consumerProvidedVariantFinder.findTransformedVariants(variants, componentRequested);
-
-        // If there are multiple potential artifact transform variants, perform attribute matching to attempt to find the best.
-        if (transformedVariants.size() > 1) {
-            transformedVariants = tryDisambiguate(matcher, transformedVariants, componentRequested);
+        ImmutableAttributes targetAttributes = attributesFactory.concat(requestAttributes, producer.getOverriddenAttributes());
+
+        // Check for matching variant without using artifact transforms.  If we found only one, return it.  If we found multiple matches, that's ambiguity.
+        List<ResolvedVariant> matchingVariants = matcher.matchMultipleCandidates(producer.getCandidates(), targetAttributes);
+        if (matchingVariants.size() == 1) {
+            return Iterables.getOnlyElement(matchingVariants).getArtifacts();
```

Comment:
Can we avoid `Iterables.getOnlyElement()` here, since `matchingVariants` is a `List`?

This code gets called very often, and we can avoid allocating an iterator each time this is called by instead calling `matches.get(0)` like we did in the prior code.

Reply:
Changed.